toolchain/parser/parser_impl.cpp

```diff
@@ -1180,6 +1180,87 @@ auto ParseTree::Parser::ParseWhileStatement() -> llvm::Optional<Node> {
                  /*has_error=*/!cond || !body);
 }
 
+auto ParseTree::Parser::ParseForStatement() -> llvm::Optional<Node> {
+  // TODO Add proper diagnostics.
+  CARBON_RETURN_IF_STACK_LIMITED(llvm::None);
+  auto start = GetSubtreeStartPosition();
+  auto token = Consume(TokenKind::For());
+
+  auto header = [this]() -> llvm::Optional<Node> {
+    auto open_paren = ConsumeIf(TokenKind::OpenParen());
+    auto header_start = GetSubtreeStartPosition();
+
+    if (!open_paren) {
+      CARBON_DIAGNOSTIC(ExpectedParenAfter, Error, "Expected `(` after `{0}`.",
+                        TokenKind);
+      emitter_.Emit(*position_, ExpectedParenAfter, TokenKind::For());
+    } else {
+      // Question: should we:
+      // 1. Try to continue parsing the loop header?
+      // 2. Skip all tokens until we reach either:
+      //    * A `)` without an opening one?
+      //      That might be easy because:
+      //      - The lexer will likely add a recovery token and mark the `)` as
+      //        an error token. So we skip until we come across this error
+      //        token.
+      //    * A `{` which marks the start of the block:
+      //      We might want to implement a heuristic here that the `{` has to be
+      //      the last token on the line.
+      //    * The end of line.
+      //
+      // The first option boils down to trying to report as many errors as
+      // possible in the header. The second one tries to skip over as much as
+      // possible until we reach a new construct, i.e. the for body block.
```

Comment:
As general guidance, erroneous parse trees are intended to be used by tools like syntax highlighters and autocomplete that want to be able to parse as much as possible even when the code is invalid, so whichever option allows us to parse the largest amount of the rest of the program is what we should choose.

It might be reasonable to scan forward until we reach one of these cases:
- An `in` token. In this case, we could recover by assuming the `(` was omitted. If after parsing the `PATTERN in EXPRESSION` we find an erroneous unmatched `)`, then skip it too.
- A closing bracket. In this case, there's probably not much we can do other than consume all the tokens to the bracket and return to the caller.
- A token at the start of a line whose indent is <= that of the line containing the `for` token. In this case, it's probably reasonable to assume that all the tokens from `for` up to but excluding that token were supposed to be part of the loop.

Reply:
See `SkipPastLikelyEnd` for skipping to a suitable recovery line / token. You may need to extend that so you can tell it to also stop at an `in` token.

Reply:
Thanks a lot for the guidance.

What suggested makes sense. However, I guess implementing it might not be possible with the current lexer recovery strategy. In particular, consider the following snippet:

```
fn foo() {
  for var x:i32 in y)
  {
    Print(x);
  }
}
```

This produces the following token buffer:
```
token: { index:  0, kind:                 'Fn', line:  1, column:  1, indent: 1, spelling: 'fn', has_trailing_space: true }
token: { index:  1, kind:         'Identifier', line:  1, column:  4, indent: 1, spelling: 'foo', identifier: 0 }
token: { index:  2, kind:          'OpenParen', line:  1, column:  7, indent: 1, spelling: '(', closing_token: 3 }
token: { index:  3, kind:         'CloseParen', line:  1, column:  8, indent: 1, spelling: ')', opening_token: 2, has_trailing_space: true }
token: { index:  4, kind:     'OpenCurlyBrace', line:  1, column: 10, indent: 1, spelling: '{', closing_token: 12, has_trailing_space: true }
token: { index:  5, kind:                'For', line:  7, column:  1, indent: 1, spelling: 'for', has_trailing_space: true }
token: { index:  6, kind:                'Var', line:  7, column:  5, indent: 1, spelling: 'var', has_trailing_space: true }
token: { index:  7, kind:         'Identifier', line:  7, column:  9, indent: 1, spelling: 'x', identifier: 1 }
token: { index:  8, kind:              'Colon', line:  7, column: 10, indent: 1, spelling: ':' }
token: { index:  9, kind: 'IntegerTypeLiteral', line:  7, column: 11, indent: 1, spelling: 'i32', has_trailing_space: true }
token: { index: 10, kind:                 'In', line:  7, column: 15, indent: 1, spelling: 'in', has_trailing_space: true }
token: { index: 11, kind:         'Identifier', line:  7, column: 18, indent: 1, spelling: 'y', identifier: 2 }
token: { index: 12, kind:    'CloseCurlyBrace', line:  7, column: 19, indent: 1, spelling: '}', opening_token: 4, recovery: true }
token: { index: 13, kind:              'Error', line:  7, column: 19, indent: 1, spelling: ')', has_trailing_space: true }
token: { index: 14, kind:     'OpenCurlyBrace', line:  8, column:  1, indent: 1, spelling: '{', closing_token: 20, has_trailing_space: true }
token: { index: 15, kind:         'Identifier', line:  9, column:  3, indent: 3, spelling: 'Print', identifier: 3 }
token: { index: 16, kind:          'OpenParen', line:  9, column:  8, indent: 3, spelling: '(', closing_token: 18 }
token: { index: 17, kind:         'Identifier', line:  9, column:  9, indent: 3, spelling: 'x', identifier: 1 }
token: { index: 18, kind:         'CloseParen', line:  9, column: 10, indent: 3, spelling: ')', opening_token: 16 }
token: { index: 19, kind:               'Semi', line:  9, column: 11, indent: 3, spelling: ';', has_trailing_space: true }
token: { index: 20, kind:    'CloseCurlyBrace', line: 10, column:  1, indent: 1, spelling: '}', opening_token: 14, has_trailing_space: true }
token: { index: 21, kind:              'Error', line: 19, column:  1, indent: 1, spelling: '}', has_trailing_space: true }
token: { index: 22, kind:          'EndOfFile', line: 19, column:  2, indent: 1, spelling: '' }
```

Token `12` above is a recovery token because with the current strategy, the lexer finds a `)` while there was a `{` that is still not closed. So it eagerly assumes that the `)` was meant to be the `}` closing `foo`'s block. In turn, this results marking token `21` (the actual `}` for `foo` as error).

So, I think a better strategy might be needed here. One that comes to mind is: search backwards until the lexer finds an introducer keyword like `var`, `for`, `while`, etc. and insert the recovery token *before* that introducer token. What do you think?

Reply:
Thinking about this a bit more; what I proposed above for the backward scan strategy doesn't help much. 

The general issue with injecting closing tokens in the buffer is that these recovering tokens might end up being inserted in the middle of a construct that doesn't have anything to do with the unclosed open token like the example above.

I am wondering if we can do without recovery tokens but that doesn't seem to be very easy since we need them to skip ahead during parsing if needed.

I will think about this issue a bit more and try out a few dumb ideas :D. Sorry for the noise.

Reply:
[As discussed on discord](https://discord.com/channels/655572317891461132/655578254970716160/1022866041791053925), we need to fix the token recovery insertion algorithm first. I added a `TODO` here to handle missing `(` or `)`. I will revisit this `TODO` once this is done.